proposals/p2188.md

```diff
@@ -0,0 +1,456 @@
+-   _field-pattern_ ::= _proper-field-pattern_
+
+A struct pattern resembles a struct literal, with at least one field initialized
+with a proper pattern:
+
+```
+match ({.a = 1, .b = 2}) {
+  case {.b = n: i32, .a = m: i32} => {}
+}
+```
+
+The scrutinee is required to be of struct type, and to have the same set of
+field names as the pattern. The pattern is matched left-to-right, meaning that
+matching is performed in the field order specified in the pattern, not in the
+field order of the scrutinee. This is consistent with the behavior of matching
+against a tuple-valued expression, where the left operand determines the order
+in which `==` comparisons are performed.
+
+In the case where a field will be bound to an identifier with the same name, a
+shorthand syntax is available: `a: T` is synonymous with `.a = a: T`.
+
+```
+match ({.a = 1, .b = 2}) {
+  case {a: i32, b: i32} => { return a + b; }
+}
+```
+
+If some fields should be ignored when matching, a trailing `, ...` can be added
+to specify this:
+
+```
+match ({.a = 1, .b = 2}) {
+  case {.a = 1, ...} => { return 1; }
+  case {b: i32, ...} => { return b; }
+}
+```
+
+This is valid even if all fields are actually named in the pattern.
+
+### Choice patterns
+
+A choice pattern is used to match one alternative of a choice type.
+
+_proper-pattern_ ::= _callee-expression_ _tuple-pattern_ _proper-pattern_ ::=
+_designator_ _tuple-pattern_?
+
+Here, _callee-expression_ is syntactically an expression that is valid as the
+callee in a function call expression, and a choice pattern is syntactically a
+function call expression whose argument list contains at least one
+_proper-pattern_.
+
+If a _callee-expression_ is provided, it is required to name a choice type
+alternative that has a parameter list, and the scrutinee is implicitly converted
+to that choice type. Otherwise, the scrutinee is required to be of some choice
+type, and the designator is looked up in that type and is required to name an
+alternative with a parameter list if and only if a _tuple-pattern_ is specified.
+
+The pattern matches if the active alternative in the scrutinee is the specified
+alternative, and the arguments of the alternative match the given tuple pattern
+(if any).
+
+```
+choice Optional(T:! Type) {
+  None,
+  Some(T)
+}
+
+match (Optional(i32).None) {
+  // ✅ `.None` resolved to `Optional(i32).None`.
+  case .None => {}
+  // ✅ `.Some` resolved to `Optional(i32).Some`.
+  case .Some(n: i32) => { Print("{0}", n); }
+  // ❌ Error, no such alternative exists.
+  case .Other => {}
+}
+
+class X {
+  external impl as ImplicitAs(Optional(i32));
+}
+
+match ({} as X) {
+  // ✅ OK, but expression pattern.
+  case Optional(i32).None => {}
+  // ✅ OK, implicitly converts to `Optional(i32)`.
+  case Optional(i32).Some(n: i32) => { Print("{0}", n); }
+}
+```
+
+Note that a pattern of the form `Optional(T).None` is an expression pattern and
+is compared using `==`.
+
+### Templates
```

Comment:
We're going to need to have a conversation about whether this should also be allowed for values with checked/symbolic generic types.

Reply:
Acknowledged. For now I want to reject, given that dispatch based on the actual type seems very template-y.